Code/Framework/AzCore/AzCore/Memory/OSAllocator.cpp

```diff
@@ -67,39 +69,42 @@ namespace AZ
             OnOutOfMemory(byteSize, alignment);
         }
 
+        size_type allocatedSize{};
 #if defined(AZ_ENABLE_TRACING)
         // We assume 1 alignment because alignment is sometimes not passed in deallocate. This does mean that we are under-reporting
         // for cases where alignment != 1 and the OS could not find a block specifically for that alignment (the OS will give use a
         // block that is byteSize+(alignment-1) and place the ptr in the first address that satisfies the alignment).
-        const size_type allocatedSize = get_allocated_size(address, 1);
+        allocatedSize = get_allocated_size(address, 1);
         m_numAllocatedBytes += allocatedSize;
         AZ_PROFILE_MEMORY_ALLOC_EX(MemoryReserved, fileName, lineNum, address, byteSize, name);
         AZ_MEMORY_PROFILE(ProfileAllocation(address, byteSize, alignment, 1));
 #endif
 
-        return address;
+        return AllocateAddress{ address, allocatedSize };
     }
 
     //=========================================================================
     // DeAllocate
     // [9/2/2009]
     //=========================================================================
-    void OSAllocator::deallocate(pointer ptr, [[maybe_unused]] size_type byteSize, [[maybe_unused]] size_type alignment)
+    auto OSAllocator::deallocate(pointer ptr, [[maybe_unused]] size_type byteSize, [[maybe_unused]] size_type alignment)
+        -> size_type
     {
+        size_type allocatedSize{};
 #if defined(AZ_ENABLE_TRACING)
         if (ptr)
         {
-            const size_type allocatedSize = get_allocated_size(ptr, 1);
+            allocatedSize = get_allocated_size(ptr, 1);
             m_numAllocatedBytes -= allocatedSize;
             AZ_PROFILE_MEMORY_FREE(MemoryReserved, ptr);
             AZ_MEMORY_PROFILE(ProfileDeallocation(ptr, byteSize, alignment, nullptr));
         }
 #endif
         AZ_OS_FREE(ptr);
-
+        return allocatedSize;
     }
 
-    OSAllocator::pointer OSAllocator::reallocate(pointer ptr, size_type newSize, align_type alignment)
+    AllocateAddress OSAllocator::reallocate(pointer ptr, size_type newSize, align_type alignment)
```

Comment:
I'm not sure I understand why some of the function headers have been changed to use `auto <function def> -> <return type>` and some haven't.  What is the goal with those changes?

Reply:
The trailing return type is used in the case when the return type is a dependent name of the allocator case.
For example the `size_type` type alias is a dependent name of the IAllocator, so by using the trailing return type, the `size_type` alias is in scope of the class and therefore is accessible without needing to specify the class twice.
So for the OSAllocator deallocate function instead of stating `OSAllocator::size_type OSAllocator::deallocate(...)` for the function signature, we use `auto OSAllocator::deallocate(...) -> size_type`.

Now the signature on `OSAllocator::allocate` has no reason to use a trailing return type as that was mistake.